specs/makeStyles.md

```diff
@@ -0,0 +1,290 @@
+Basic components still use Fela with atomic CSS classes and 1 class specificity. `useCSS` is used for overrides and always generates classes with 2 classes specificity. Also, instead of using atomic CSS, monolithic classes are used for overrides (Emotion CSS-in-JS is used for that).
+
+#### No special prop for passing overrides to children
+
+For handling multiple levels of overrides, instead of using a special prop with all its problems (see above), `useCSS` just returns a class which can be passed to a child using `className`. If there is another `useCSS` call in the child, it takes the `className` from parent.
+
+1. `useCSS` call in parent generates a classname, and adds mapping from the classname to a style object to a global dictionary.
+2. Parent takes the class returned by the `useCSS` call and passes it to its child.
+3. The child calls `useCSS` to apply its own overrides and also passes it the class from parent.
+4. `useCSS` finds the style object in the dictionary, merges the original style object with the child's style overrides and returns the new classname. Instead of performing an expensive object deep merge, it just concatenates the two stringified style objects and depends on the-latest-wins CSS rule.
+
+![](makeStyles-useCss.png)
+
+#### Limitations
+
+This approach looks good for style overrides, we are not sure this approach works the best for core component styles. Without that, we would have two different styling approaches in place.
+Although the object string concatenation is performant, it results in a lot of CSS code.
+
+### v8 `makeStyles`
+
+In order to improve rendering performance, v8 approach is closer to static CSS.
+
+There are two main functions in this approach `makeStyles` and `useStyles`.
+`makeStyles` is a factory which takes style object as input and returns a `useStyles` hook. `makeStyles` does not accept any theme or React context and should be called in a module scope, not in a scope of a React component.
+When rendering a component, `useStyles` hook is called passing a theme to it. The `useStyles` function injects the styles into DOM (if not done yet) and maps props to classnames.
+
+#### Multiple classes are applied to a single element
+
+As in static CSS approach, multiple semantic classes are applied to an element.
+For example `<Button primary disabled>` renders as `<button class="primary disabled">`.
+To style the component, combined classes are used in CSS as well:
+
+```css
+.primary {
+}
+.disabled {
+}
+.primary.disabled {
+}
+```
+
+This is causing problems with overrides - assuming it is guaranteed that override classes are written bellow the component styles in the DOM, for most cases in this example, it is enough to have 1 class specificity for overrides to work correctly. The only case where the 1 class override does not work is for CSS properties in `.disabled.button`.
+Consumer does not know what specificity must be used for the overrides, this can also change between library versions.
```

Comment:
Perhaps that 2 different engineers can break each other's styling ? i.e. the specificity wars